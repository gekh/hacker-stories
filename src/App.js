import React from "react";

const initialStories = [
  {
    title: 'React',
    url: 'https://reactjs.org',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
  }
];

const getAsyncStories = () =>
  new Promise((resolve) => setTimeout(
    () => resolve({ data: { stories: initialStories } }),
    200
  ));

const useSemipersistentState = (key, initialState) => {
  const [value, setValue] = React.useState(localStorage.getItem('search') || initialState);

  React.useEffect(() => {
    localStorage.setItem(key, value);
  }, [key, value]);

  return [value, setValue];
};

const App = () => {

  const [searchTerm, setSearchTerm] = useSemipersistentState('search', 'React');
  const [stories, setStories] = React.useState([]);

  React.useEffect(() => {
    getAsyncStories().then(result => {
      setStories(result.data.stories);
    });
  }, []);

  const handleRemoveStory = (item) => {
    const filteredStories = stories.filter((story) => item.objectID !== story.objectID);
    setStories(filteredStories);
  };

  const handleChange = event => {
    setSearchTerm(event.target.value);
  }

  const searchedStories = stories.filter(story =>
    story.title.toLowerCase().includes(searchTerm.toLowerCase())
  );

  return (
    <div>
      <h1>My Hacker Stories</h1>

      <InputWithLabel id="search" onChange={handleChange} value={searchTerm} isFocused>
        Search:
      </InputWithLabel>

      you're searching for: {searchTerm}

      <hr />

      <List list={searchedStories} onRemoveItem={handleRemoveStory} />

    </div>
  );
};

const List = ({ list, onRemoveItem }) => (
  <ul>
    {list.map(item =>
      <Item item={item} onRemoveItem={onRemoveItem} />
    )}
  </ul>
);

const Item = ({ item, onRemoveItem }) => (
  <li key={item.objectID}>
    <span>
      <a href={item.url}>{item.title}</a>
    </span>
    {' '}—{' '}
    <span>{item.author}</span>
    {' '}—{' '}
    <span>{item.num_comments}</span>
    {' '}—{' '}
    <span>{item.points}</span>
    {' '}—{' '}
    <span>
      <button type="button" onClick={onRemoveItem.bind(null, item)}>
        Dismiss
      </button>
    </span>
  </li>
);

const InputWithLabel = ({ id, type = 'text', value, onChange, isFocused, children }) => {

  const inputRef = React.useRef();

  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <>
      <label htmlFor={id}>{children}</label>
      &nbsp;
      <input id={id} type={type} value={value} onChange={onChange} ref={inputRef} />
    </>
  );
}

export default App;
