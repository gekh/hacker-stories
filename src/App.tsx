import React from 'react';
import axios from 'axios';
import styled from 'styled-components';
import {ReactComponent as Check} from './check.svg';

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';
const STORIES_FETCH_INIT = 'STORIES_FETCH_INIT';
const STORIES_FETCH_SUCCESS = 'STORIES_FETCH_SUCCESS';
const STORIES_FETCH_FAILURE = 'STORIES_FETCH_FAILURE';
const REMOVE_STORY = 'REMOVE_STORY';

type Story = {
	objectID: string;
	url: string;
	title: string;
	author: string;
	num_comments: number;
	points: number;
};

type Stories = Story[];

type ItemProps = {
	item: Story;
	onRemoveItem: (item: Story) => void;
};

type ListProps = {
	list: Stories;
	onRemoveItem: (item: Story) => void;
};

type StoriesState = {
	data: Stories;
	isLoading: boolean;
	isError: boolean;
};

type StoriesFetchInitAction = {
	type: 'STORIES_FETCH_INIT';
};

type StoriesFetchSuccessAction = {
	type: 'STORIES_FETCH_SUCCESS';
	payload: Stories;
};

type StoriesFetchFailurAction = {
	type: 'STORIES_FETCH_FAILURE';
};

type StoriesRemoveAction = {
	type: 'REMOVE_STORY';
	payload: Story;
};

type StoriesAction =
  StoriesFetchInitAction
  | StoriesFetchSuccessAction
  | StoriesFetchFailurAction
  | StoriesRemoveAction;

type SearchFormProps = {
	searchTerm: string;
	onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
	onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
};

type InputWithLabelProps = {
	id: string;
	value: string;
	type?: string;
	onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
	isFocused?: boolean;
	children: React.ReactNode;
};

const storiesReducer = (state: StoriesState, action: StoriesAction) => {
	switch (action.type) {
		case STORIES_FETCH_INIT:
			return {
				...state,
				isLoading: true,
				isError: false,
			};

		case STORIES_FETCH_SUCCESS:
			return {
				...state,
				isLoading: false,
				isError: false,
				data: action.payload,
			};

		case STORIES_FETCH_FAILURE:
			return {
				...state,
				isLoading: true,
				isError: false,
			};

		case REMOVE_STORY:
			return {
				...state,
				data: state.data.filter(story => action.payload.objectID !== story.objectID),
			};

		default:
			throw new Error();
	}
};

const useSemipersistentState = (
	key: string,
	initialState: string,
): [string, (newValue: string) => void] => {
	const [value, setValue] = React.useState(localStorage.getItem('search') || initialState);

	React.useEffect(() => {
		localStorage.setItem(key, value);
	}, [key, value]);

	return [value, setValue];
};

const StyledContainer = styled.div`
  height: 100vw;
  padding: 20px;

  background: #83a4d4;
  background: linear-gradient(to left, #b6fbff, #83a4d4);

  color: #171212;
`;

const StyledHeadlinePrimary = styled.h1`
  font-size: 48px;
  font-weight: 300;
  letter-spacing: 2px;
`;

const getSumComments = (stories: {data: Story[]}) => stories.data.reduce(
	(result: number, value: Story) => result + value.num_comments,
	0,
);

const App = () => {
	const [searchTerm, setSearchTerm] = useSemipersistentState('search', 'React');

	const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

	const [stories, dispatchStories] = React.useReducer(storiesReducer, {
		data: [],
		isLoading: false,
		isError: false,
	});

	const handleFetchStories = React.useCallback(async () => {
		dispatchStories({type: STORIES_FETCH_INIT});

		try {
			const result = await axios.get(url);

			dispatchStories({
				type: STORIES_FETCH_SUCCESS,
				payload: result.data.hits,
			});
		} catch {
			dispatchStories({
				type: STORIES_FETCH_FAILURE,
			});
		}
	}, [url]);

	React.useEffect(() => {
		handleFetchStories();
	}, [handleFetchStories]);

	const handleRemoveStory = React.useCallback((item: Story) => {
		dispatchStories({
			type: REMOVE_STORY,
			payload: item,
		});
	}, []);

	const handleSearchInput = React.useCallback((
		event: React.ChangeEvent<HTMLInputElement>,
	) => {
		setSearchTerm(event.target.value);
	}, [setSearchTerm]);

	const handleSearchSubmit = React.useCallback((
		event: React.FormEvent<HTMLInputElement>,
	) => {
		setUrl(`${API_ENDPOINT}${searchTerm}`);

		event.preventDefault();
	}, [searchTerm]);

	const sumComments = React.useMemo(() => getSumComments(stories), [stories]);

	// What's going on?
	//! wow easy maaan
	// TODO update the funciton
	return (
		<StyledContainer>
			<StyledHeadlinePrimary>My Hacker Stories with {sumComments} comments</StyledHeadlinePrimary>

			<SearchForm
				onSearchSubmit={handleSearchSubmit}
				onSearchInput={handleSearchInput}
				searchTerm={searchTerm}
			></SearchForm>

			{stories.isError && <p>Something went wrong...</p>}

			{stories.isLoading
				? (
					<p> Is loading...</p>
				)
				: (
					<List list={stories.data} onRemoveItem={handleRemoveStory} />
				)}
		</StyledContainer>
	);
};

const List = React.memo(({list, onRemoveItem}: ListProps) => (
	<ul>
		{list.map(item =>
			<Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />,
		)}
	</ul>
));

const StyledItem = styled.li`
  display: flex;
  align-items: center;
  padding-bottom: 5px;
`;

const StyledColumn = styled.span`
  padding: 0 5px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  a {
    color: inherit;
  }

  width: ${props => props.width};
`;

const StyledButton = styled.button`
  background: transparent;
  border: 1px solid #171212;
  padding: 5px;
  cursor: pointer;
  margin: 0 5px;

  transition: all 0.1s ease-in;

  &:hover {
    background: #171212;
    color: #ffffff;
    fill: #ffffff;
    stroke: #ffffff;
  }
`;

const StyledButtonSmall = styled(StyledButton)`
  padding: 5px;
`;

const StyledButtonLarge = styled(StyledButton)`
  padding: 10px;
`;

const Item = ({item, onRemoveItem}: ItemProps) => (
	<StyledItem>
		<StyledColumn width='40%'>
			<a href={item.url}>{item.title}</a>
		</StyledColumn>
		<StyledColumn width='30%'>{item.author}</StyledColumn>
		<StyledColumn width='10%'>{item.num_comments}</StyledColumn>
		<StyledColumn width='10%'>{item.points}</StyledColumn>
		<StyledColumn width='10%'>
			<StyledButtonSmall
				type='button'
				onClick={onRemoveItem.bind(null, item)}
			>
				<Check height='18px' width='18px' />
			</StyledButtonSmall>
		</StyledColumn>
	</StyledItem>
);

const StyledLabel = styled.label`
  border-top: 1px solid #171212;
  border-left: 1px solid #171212;
  padding-left: 5px;
  font-size: 24px;
`;

const StyledInput = styled.input`
  border: none;
  border-bottom: 1px solid #171212;
  background-color: transparent;

  margin: 0 5px;
  
  font-size: 24px;
`;

const InputWithLabel = ({
	id,
	type = 'text',
	value,
	onInputChange,
	isFocused,
	children,
}: InputWithLabelProps) => {
	const inputRef = React.useRef<HTMLInputElement>(null);

	React.useEffect(() => {
		if (isFocused && (inputRef.current != null)) {
			inputRef.current.focus();
		}
	}, [isFocused]);

	return (
		<>
			<StyledLabel htmlFor={id}>{children}</StyledLabel>
      &nbsp;
			<StyledInput
				id={id}
				type={type}
				value={value}
				onChange={onInputChange}
				ref={inputRef}
			/>
		</>
	);
};

const StyledSearchForm = styled.form`
  padding:  10px 0 20px 0;
  display: flex;
  align-items: baseline;
`;

const SearchForm = React.memo(({
	onSearchSubmit,
	onSearchInput,
	searchTerm,
}: SearchFormProps) => console.log('SearchForm') || (
	<StyledSearchForm onSubmit={handleSearchSubmit}>
		<InputWithLabel
			id='search'
			value={searchTerm}
			isFocused
			onInputChange={handleSearchInput}
		>
      Search:
		</InputWithLabel>

		<StyledButtonLarge
			type='submit'
			disabled={!searchTerm}
		>
      Submit
		</StyledButtonLarge>
	</StyledSearchForm>
));

export default App;
