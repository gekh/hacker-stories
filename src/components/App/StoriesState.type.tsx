import { Stories } from "../../types/stories.type";

type StoriesState = {
  data: Stories;
  isLoading: boolean;
  isError: boolean;
};

export type { StoriesState };
