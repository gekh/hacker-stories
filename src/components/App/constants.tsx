export const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';
export const STORIES_FETCH_INIT = 'STORIES_FETCH_INIT';
export const STORIES_FETCH_SUCCESS = 'STORIES_FETCH_SUCCESS';
export const STORIES_FETCH_FAILURE = 'STORIES_FETCH_FAILURE';
export const REMOVE_STORY = 'REMOVE_STORY';