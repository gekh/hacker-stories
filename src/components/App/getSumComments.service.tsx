import { Story } from "../../types/stories.type";

export const getSumComments = (stories: { data: Story[]; }) =>
  stories.data.reduce(
    (result: number, value: Story) => result + value.num_comments,
    0
  );
