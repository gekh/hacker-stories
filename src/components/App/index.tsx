import axios from 'axios';
import React, { ChangeEvent, FormEvent } from 'react';
import { Story } from '../../types/stories.type';
import { List } from '../List';
import { SearchForm } from '../SearchForm';
import {
  API_ENDPOINT, REMOVE_STORY, STORIES_FETCH_FAILURE, STORIES_FETCH_INIT,
  STORIES_FETCH_SUCCESS
} from './constants';
import { getSumComments } from './getSumComments.service';
import { storiesReducer } from './stories.reducer';
import { StyledContainer, StyledHeadlinePrimary } from './styles';
import { useSemipersistentState } from './useSemipersistentState.hook';


const App = () => {
  const [searchTerm, setSearchTerm] = useSemipersistentState('search', 'React');

  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  const handleFetchStories = React.useCallback(async () => {
    dispatchStories({ type: STORIES_FETCH_INIT });

    try {
      const result = await axios.get(url);

      dispatchStories({
        type: STORIES_FETCH_SUCCESS,
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({
        type: STORIES_FETCH_FAILURE,
      });
    }
  }, [url]);

  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = React.useCallback((item: Story) => {
    dispatchStories({
      type: REMOVE_STORY,
      payload: item,
    });
  }, []);

  const handleSearchInput = React.useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      setSearchTerm(event.target.value);
    },
    [setSearchTerm]
  );

  const handleSearchSubmit = React.useCallback(
    (event: FormEvent<HTMLFormElement>) => {
      setUrl(`${API_ENDPOINT}${searchTerm}`);

      event.preventDefault();
    },
    [searchTerm]
  );

  const sumComments = React.useMemo(() => getSumComments(stories), [stories]);

  // What's going on?
  //! wow easy maaan
  // TODO: update the funciton
  return (
    <StyledContainer>
      <StyledHeadlinePrimary>
        My Hacker Stories with {sumComments} comments
      </StyledHeadlinePrimary>

      <SearchForm
        onSearchSubmit={handleSearchSubmit}
        onSearchInput={handleSearchInput}
        searchTerm={searchTerm}
      ></SearchForm>

      {stories.isError && <p>Something went wrong...</p>}

      {stories.isLoading ? (
        <p> Is loading...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </StyledContainer>
  );
};

export default App;
export { storiesReducer };
