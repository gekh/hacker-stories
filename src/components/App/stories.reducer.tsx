import { StoriesAction } from '../../types/stories-acton.type';
import {
  STORIES_FETCH_INIT,
  STORIES_FETCH_SUCCESS,
  STORIES_FETCH_FAILURE,
  REMOVE_STORY
} from './constants';
import { StoriesState } from './StoriesState.type';


export const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch (action.type) {
    case STORIES_FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: false,
      };

    case STORIES_FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };

    case STORIES_FETCH_FAILURE:
      return {
        ...state,
        isLoading: false,
        isError: true,
      };

    case REMOVE_STORY:
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };

    default:
      throw new Error();
  }
};
