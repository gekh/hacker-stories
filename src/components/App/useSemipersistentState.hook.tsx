import React from 'react';


export const useSemipersistentState = (
  key: string,
  initialState: string
): [string, (newValue: string) => void] => {
  const [value, setValue] = React.useState(
    localStorage.getItem('search') || initialState
  );

  React.useEffect(() => {
    localStorage.setItem(key, value);
  }, [key, value]);

  return [value, setValue];
};
