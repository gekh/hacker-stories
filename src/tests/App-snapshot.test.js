import { render, screen, fireEvent } from '@testing-library/react';
import { storiesReducer, SearchForm, InputWithLabel, List, Item } from '../App';
import '@testing-library/jest-dom';
import { storyOne, storyTwo, stories } from './fixtures/stories.fixture';


describe('Item', () => {
  it('renders a snapshot', () => {
    const { container } = render(<Item item={storyOne} onRemoveItem={jest.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('SearchForm', () => {
  const searchFormProps = {
    searchTerm: 'React',
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn().mockImplementation(e => e.preventDefault()),
  };

  it('renders a snapshot', () => {
    const { container } = render(<SearchForm {...searchFormProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('InputWithLabel', () => {
  it('renders a snapshot', () => {
    const { container } = render(<InputWithLabel
      id="search"
      value="React"
      onInputChange={jest.fn()}
    >
      Search:
    </InputWithLabel>);

    expect(container.firstChild).toMatchSnapshot();
  });
});
