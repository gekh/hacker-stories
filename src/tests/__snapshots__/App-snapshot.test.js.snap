// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputWithLabel renders a snapshot 1`] = `
<label
  class="sc-iBYQkv dmSATp"
  for="search"
>
  Search:
</label>
`;

exports[`Item renders a snapshot 1`] = `
<li
  class="sc-dkrFOg bTWlcS"
>
  <span
    class="sc-hLBbgP cqbfSs"
    width="40%"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    class="sc-hLBbgP cqbSZz"
    width="30%"
  >
    Jordan Walke
  </span>
  <span
    class="sc-hLBbgP cqbtYF"
    width="10%"
  >
    3
  </span>
  <span
    class="sc-hLBbgP cqbtYF"
    width="10%"
  >
    4
  </span>
  <span
    class="sc-hLBbgP cqbtYF"
    width="10%"
  >
    <button
      class="sc-eDvSVe sc-jSUZER jeFzlQ dEKhgm"
      type="button"
    >
      Dismiss
    </button>
  </span>
</li>
`;

exports[`SearchForm renders a snapshot 1`] = `
<form
  class="sc-pyfCe jNZiVM"
>
  <label
    class="sc-iBYQkv dmSATp"
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    class="sc-ftTHYK gVUtwk"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="sc-eDvSVe sc-gKPRtg jeFzlQ eGWFJW"
    type="submit"
  >
    Submit
  </button>
</form>
`;
